contracts/credits/src/msg.rs

```diff
@@ -0,0 +1,154 @@
+use crate::state::Allocation;
+use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::{Addr, Timestamp, Uint128};
+
+#[cw_serde]
+pub struct InstantiateMsg {
+    /// DAO contract address
+    pub dao_address: String,
+    /// Airdrop contract address
+    pub airdrop_address: Option<String>,
+    /// Lockdrop contract address,
+    pub lockdrop_address: Option<String>,
+    /// When can start withdrawing NTRN funds
+    pub when_withdrawable: Timestamp,
+}
+
+#[cw_serde]
+pub enum ExecuteMsg {
+    /// UpdateConfig is a message to initialize all addresses.
+    /// Needed because there are circle deps between contracts.
+    /// [Permissioned - DAO]
+    UpdateConfig {
+        airdrop_address: String,
+        lockdrop_address: String,
+    },
+    /// AddVesting is a message that allows address to claim particular amount of NTRNs at particular time.
+    /// Can only store one vesting amount per address.
+    /// [Permissioned - Airdrop address]
+    AddVesting {
+        address: String,
+        amount: Uint128,
+        start_time: u64,
+        duration: u64,
+    },
+    /// Transfer is a base message to move tokens to another account without triggering actions.
+    /// [Permissioned - Airdrop address]
+    Transfer { recipient: String, amount: Uint128 },
+    /// Withdraw is a message that burns all vested CNTRN tokens
+    /// on the sender and sends NTRN tokens in 1:1 proportion.
+    /// [Permissionless]
+    Withdraw {},
+    /// Burn is a message only for airdrop account to destroy
+    /// certain amount of CNTRN's forever and send NTRN tokens in 1:1 proportion.
+    /// [Permissioned - Airdrop address]
+    Burn { amount: Uint128 },
+    /// BurnFrom is a message only for lockdrop contract
+    /// to burn owner's CNTRN tokens and mint NTRN tokens in 1:1 proportion certain amount for owner.
+    /// Used to skip vesting as a reward for participating in the lockdrop.
+    /// [Permissioned - Lockdrop address]
+    BurnFrom { owner: String, amount: Uint128 },
+    /// If authorized (only dao can call),
+    /// locks the NTRN tokens and mints CNTRN tokens in 1:1 amount
+    /// and adds to the dao balance.
+    /// [Permissioned - DAO] (Dao set in initialize func as cw20 minter)
+    Mint {},
+}
+
+#[cw_serde]
+#[derive(QueryResponses)]
```

Comment:
these `#[cw_serde]` and `#[derive(QueryResponses)]` looks nice, but I still never know how to understand these macro things in rust. what do these macros do?

Reply:
cw_serde is basically replaces all those json decoder and encoder directives
derive(QueryResponses) mostly about specifying what response we expect from what enum (don't know the purpose of that exactly)